### ISSUE TITLE
[Feature Request] Keep delta_t as floats

### ISSUE THREAD

**Issue by Expertium** [open-spaced-repetition/fsrs4anki#437]
Other day I was thinking that it might be worth to keep review intervals as floats, as opposed to calculating the day boundary and rounding them to days. Apart from helping those who travel a lot, it will remove necessity to enter timezone and day start, which is annoying an error prone, and also, it might improve accuracy for short interval reviews. Common sense suggests that if I learned a new card in evening and then reviewed it next day morning, I'll have better chance of remembering it if I was to learn it in morning and review it next day evening, however to the optimizer these two intervals are the same. I think this is also related to the spike on the stability to retention graph, which is at around 0.2 in my case. Despite this, I don't know if this change will bring any practical improvement.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/137687989/db3eb37e-2c24-4e9f-b706-b2fe8c89b354)

_Originally posted by @nb9618 in https://github.com/open-spaced-repetition/fsrs4anki/issues/429#issuecomment-1684936828_

**Comment by Expertium** [comment]
No, I mean `delta_t` and `t_history`. 
This obviously won't affect the scheduling as it is impossible to schedule a card at a specific time of a day, but it might improve accuracy of `w` and, if implemented in the add-on, accuracy of cards' memory states.

_Originally posted by @nb9618 in https://github.com/open-spaced-repetition/fsrs4anki/issues/429#issuecomment-1684945568_

**Comment by L-M-Sherlock** [comment]
And the scheduler also only knows the elapsed days, which is an integer.

**Comment by user1823** [comment]
> And the scheduler also only knows the elapsed days, which is an integer.

But things are changing. When the scheduler will get integrated into Anki's rust backend, several limitations would be removed. I realise that all this would not happen very soon. But, while discussing things like this one, we should think of the long term.

**Comment by dae** [comment]
I wonder whether it's really that useful, as once intervals grow above a few days, the fractional part is rather meaningless. For the training portion, revlog entries record days, so you'd need to infer the actual elapsed time based on the timestamps. For reviewing, the way the code is currently structured, only days are available to the scheduler.

**Comment by L-M-Sherlock** [comment]
| Algorithm | Log Loss | RMSE | RMSE(bins) |
| --- | --- | --- | --- |
| FSRS v4 | 0.3820 | 0.3311 | 0.0547 |
| FSRS v4 (float delta_t) | 0.3870 | 0.3333 | 0.0557 |

The performance becomes worse.

**Comment by L-M-Sherlock** [comment]
Sorry, I make a mistake in the pretrain stage. float delta_t is incompatible with the current pretrain implementation. So I round the delta_t in the pretrain stage. The final result is:

| Algorithm | Log Loss | RMSE | RMSE(bins) |
| --- | --- | --- | --- |
| FSRS v4 | 0.3820 | 0.3311 | 0.0547 |
| FSRS v4 (float delta_t) | 0.3840 | 0.3317 | 0.0542 |

Only RMSE(bins) is improved slightly.

**Comment by L-M-Sherlock** [comment]
### Weighted by number of reviews

| Algorithm | Log Loss | RMSE | RMSE(bins) |
| --- | --- | --- | --- |
| LSTM float delta_t | 0.4219 | 0.3431 | 0.0680 |
| LSTM | 0.4193 | 0.3424 | 0.0662 |

### Weighted by ln(number of reviews)

| Algorithm | Log Loss | RMSE | RMSE(bins) |
| --- | --- | --- | --- |
| LSTM float delta_t | 0.5886 | 0.3755 | 0.1387 |
| LSTM | 0.5934 | 0.3755 | 0.1382 |

It doesn't improve the performance to keep delta_t as floats. I think it's OK to close this issue.
